File: zyx/tests/mnist.py
import time
from operator import ne

import torch
from safetensors.torch import load_file
from torch import nn
from torch.nn import functional as F


def cross_entropy_one_hot(logits, targets):
    log_probs = torch.log_softmax(logits, dim=-1)
    loss = -(targets * log_probs).sum(dim=-1)
    return loss.mean()


class MnistNet(nn.Module):
    def __init__(self, state_dict):
        super().__init__()
        self.l1_weight = nn.Parameter(state_dict["l1.weight"])
        self.l1_bias = nn.Parameter(state_dict["l1.bias"])
        self.l2_weight = nn.Parameter(state_dict["l2.weight"])
        self.l2_bias = nn.Parameter(state_dict["l2.bias"])

    def forward(self, x):
        x = x.reshape([-1, 784])
        # print(f"x={x.reshape(-1, 28, 28)[0, 15:20, 15:20]}")
        # print(f"{self.l1_weight[0, 0:10]}")
        x = x.matmul(self.l1_weight.T) + self.l1_bias
        x = x.relu()
        x = x.matmul(self.l2_weight.T) + self.l2_bias
        return x


state_dict = load_file("../zyx-examples/models/mnist.safetensors")
net = MnistNet(state_dict)

state_dict = load_file("../zyx-examples/models/mnist.safetensors")

train_dataset = load_file("../zyx-examples/data/mnist_dataset.safetensors")

train_x = train_dataset["train_x"].float() / 255.0
train_y = train_dataset["train_y"].long()
test_x = train_dataset["test_x"].float() / 255.0
test_y = train_dataset["test_y"].long()

batch_size = 64
num_train = train_x.shape[0]


optimizer = torch.optim.SGD(
    net.parameters(),
    lr=0.1,
    momentum=0.6,
    nesterov=False,
)

for epoch in range(1, 6):
    total_loss = 0.0
    iters = 0

    for i in range(0, num_train, batch_size):
        end = min(i + batch_size, num_train)

        x = train_x[i:end]
        y = train_y[i:end]

        optimizer.zero_grad()

        logits = net(x)
        print(f"{logits[0:4, :4]}")
        y_one_hot = F.one_hot(y, num_classes=10).float()
        loss = cross_entropy_one_hot(logits, y_one_hot)
        print(f"loss={loss.item()}")

        loss.backward()

        print(net.l2_bias.grad)
        # tensor([-0.0065, -0.0215,  0.0218, -0.0262, -0.0322,  0.0448,  0.0106,  0.0165, 0.0262, -0.0336])

        raise Exception

        optimizer.step()

        total_loss += loss.item()
        print(f"{loss.item()}")
        time.sleep(1)

        iters += 1

File: zyx/tests/mnist.rs
// Copyright (C) 2025 zk4x
// SPDX-License-Identifier: LGPL-3.0-only

use std::collections::HashMap;

use zyx::{DType, ReduceOp, Tape, Tensor, ZyxError};

#[test]
fn mnist() -> Result<(), ZyxError> {
    struct MnistNet {
        l1_weight: Tensor,
        l1_bias: Tensor,
        l2_weight: Tensor,
        l2_bias: Tensor,
    }

    impl MnistNet {
        fn forward(&self, x: &Tensor) -> Tensor {
            let x = x.reshape([0, 784]).unwrap();
            let x = x.matmul(&self.l1_weight.t()).unwrap() + &self.l1_bias;
            let x = x.relu();
            let x = x.matmul(&self.l2_weight.t()).unwrap() + &self.l2_bias;
            x
        }
    }

    let train_dataset: HashMap<String, Tensor> = Tensor::load("../zyx-examples/data/mnist_dataset.safetensors")?;
    let train_x = train_dataset["train_x"].cast(DType::F32) / 255;
    let train_y = train_dataset["train_y"].clone();

    let batch_size = 64;
    let num_train = train_x.shape()[0];

    let net = MnistNet {
        l1_weight: Tensor::randn([128, 784], DType::F32)?,
        l1_bias: Tensor::randn([128], DType::F32)?,
        l2_weight: Tensor::randn([10, 128], DType::F32)?,
        l2_bias: Tensor::randn([10], DType::F32)?,
    };

    for _ in 0..1 {
        for i in (0..num_train as u64).step_by(batch_size) {
            let end = if i + batch_size as u64 <= num_train as u64 {
                i + batch_size as u64
            } else {
                num_train as u64
            };

            let x = train_x.slice([i..end])?;
            let y = train_y.slice([i..end])?;

            let tape = Tape::new([&net.l1_weight, &net.l1_bias, &net.l2_weight, &net.l2_bias])?;
            let logits = net.forward(&x);
            let loss = logits.cross_entropy(y, ReduceOp::Mean)?;
            let grads = tape.gradient(&loss, [&net.l1_weight, &net.l1_bias, &net.l2_weight, &net.l2_bias, &loss]);

            // Simulate SGD update
            let lr = 0.01f32;
            let new_w1 = &net.l1_weight - &grads[0] * lr;
            let new_b1 = &net.l1_bias - &grads[1] * lr;
            let new_w2 = &net.l2_weight - &grads[2] * lr;
            let new_b2 = &net.l2_bias - &grads[3] * lr;

            tape.realize([&new_w1, &new_b1, &new_w2, &new_b2, &loss])?;
            break;
        }
    }
    Ok(())
}
